src/rdkafka_broker.c

```diff
@@ -3614,20 +3650,125 @@ rd_kafka_fetch_reply_handle (rd_kafka_broker_t *rkb,
+                                         * from the Java implementation.
+                                         * 
+                                         * Aborted transaction start offsets are arranged in a map
+                                         * (ABORTED_TXN_OFFSETS), with PID as the key and value as follows:
+                                         *  - OFFSETS:  sorted list of aborted transaction start offsets (ascending)
+                                         *  - IDX:      an index into OFFSETS list, initialized to 0.
+                                         *
+                                         * The logic for processing fetched data is as follows:
+                                         * 
+                                         * 1. If the message is a transaction control message and the status is ABORT
+                                         *   then increment ABORTED_TXN_OFFSETS(PID).IDX. note: sanity check that
+                                         *   OFFSETS[ABORTED_TXN_OFFSETS(PID).IDX] is less than the current offset
+                                         *   before incrementing. If the status is COMMIT, do nothing.
+                                         * 
+                                         * 2. If the message is a normal message, find the corresponding OFFSETS list
+                                         *   in ABORTED_TXN_OFFSETS. If it doesn't exist, then keep the message. If
+                                         *   the PID does exist, compare ABORTED_TXN_OFFSETS(PID).IDX with
+                                         *   len(OFFSETS). If it's >= then the message should be kept. If not,
+                                         *   compare the message offset with OFFSETS[ABORTED_TXN_OFFSETS(PID).IDX].
+                                         *   If it's greater than or equal to this value, then the message should be
+                                         *   ignored. If it's less than, then the message should be kept.
+                                         * 
+                                         * Note: A MessageSet comprises messages from at most one transaction, so the
+                                         * logic in step 2 is done at the message set level.
+                                         */
+
+                                        for (k = 0 ; k < AbortedTxnCnt; k++) {
+                                                int64_t pid;
+                                                int64_t first_offset;
+                                                int64_t *v;
+                                                rd_kafka_aborted_txn_start_offsets_t node, *node_ptr;
+
+                                                rd_kafka_buf_read_i64(rkbuf, &pid);
+                                                rd_kafka_buf_read_i64(rkbuf, &first_offset);
```

Comment:
the other code in this function uses CamelCase for protocol field names, suggest doing the same here: FirstOffset

Reply:
ok, I didn't fully appreciate that convention. Will also change pid to Pid.